File: Excersice 1/Code/Python/er1.3.3.1.py
import matplotlib

matplotlib.use("TkAgg")
import numpy as np
import matplotlib.pyplot as plt


N = 1000  # Samples/deigmata
np.random.seed(0)
x = np.random.randn(N)  # Input signal/sima eisodou
h = np.array([1, -0.4, -4, 0.5])  # Coefficients/Sidelestes
d = np.convolve(x, h, mode='full')[:N]  # d(n)
mu = 0.01  # Step size/vima μ

# Filter length/mhkos filtrou
filter_lengths = [3, 5]


learning_curves = []

for L in filter_lengths:
    w = np.zeros(L)  # Initialize LMS filter coefficients
    mse = np.zeros(N)  # To track squared error

# Loop for each length/epanalipseis gia to kathe mhkos
# Lms implementation/ilopoihsh lms
    for n in range(L, N):

        X = x[n:n - L:-1]

        # filter output/eksodos filtrou
        y = np.dot(w, X)

        # Error of filter output signal compared to the desired signal/sfalma eksodou
        e = d[n] - y

        # New filter coefficients/Enimerosi sideleston me kathe epanalipsi
        w += mu * e * X


        mse[n] = e ** 2


    cumulative_mse = np.cumsum(mse) / np.arange(1, N + 1)
    learning_curves.append(cumulative_mse)


plt.figure(figsize=(10, 6))
for idx, L in enumerate(filter_lengths):
    plt.plot(learning_curves[idx], label=f'L = {L}')
plt.title('Learning Curves for LMS Algorithm')
plt.xlabel('Iterations')
plt.ylabel('Mean Squared Error (MSE)')
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()  

File: Exercise 1/Code/Matlab/er_1_3_3_1.m

N = 1000; % Samples/deigmata
x = randn(1, N); % Input signal/sima eisodou
h = [1, -0.4, -4, 0.5]; % Coefficients/Sidelestes
d = filter(h, 1, x); % d(n)
mu = 0.01; % Step size/vima μ

% Filter length/mhkos filtrou
filter_lengths = [3, 5];


learning_curves = zeros(length(filter_lengths), N);

% Loop for each length/epanalipseis gia to kathe mhkos
for idx = 1:length(filter_lengths)
    L = filter_lengths(idx); 
    w = zeros(1, L); 
    mse = zeros(1, N); 

    % Lms implementation/ilopoihsh lms
    for n = L:N
       
        X = x(n:-1:n-L+1); 
        
        % filter output/eksodos filtrou
        y = w * X'; 
        
       % Error of filter output signal compared to the desired signal/sfalma
    % eksodou
        e = d(n) - y; 
        
        %  New filter coefficients/Enimerosi sideleston me kathe epanalipsi
        w = w + mu * e * X; 
        
        
        mse(n) = e^2;
    end

   
    learning_curves(idx, :) = cumsum(mse) ./ (1:N);
end


figure;
for idx = 1:length(filter_lengths)
    plot(1:N, learning_curves(idx, :), 'LineWidth', 1.5); hold on;
end
title('Learning Curves');
xlabel('Iterations');
ylabel('Mean Squared Error (MSE)');
legend(arrayfun(@(L) sprintf('L = %d', L), filter_lengths, 'UniformOutput', false));
grid on;
